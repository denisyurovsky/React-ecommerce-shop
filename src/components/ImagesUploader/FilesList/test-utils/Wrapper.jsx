import PropTypes from 'prop-types';
import React from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';

import { FilesList } from '../FilesList';

const Wrapper = ({ isLoad, files, setCroppedImage, handleDelete }) => (
  <DndProvider backend={HTML5Backend}>
    <FilesList
      handleImageDelete={handleDelete}
      filesList={files}
      setCroppedImage={setCroppedImage}
      isImagesLoad={isLoad}
    />
  </DndProvider>
);

Wrapper.defaultProps = {
  isLoad: false,
  files: [],
  setCroppedImage: () => {},
  handleDelete: () => {},
};

Wrapper.propTypes = {
  isLoad: PropTypes.bool,
  files: PropTypes.arrayOf(PropTypes.object),
  setCroppedImage: PropTypes.func,
  handleDelete: PropTypes.func,
};

export default Wrapper;
