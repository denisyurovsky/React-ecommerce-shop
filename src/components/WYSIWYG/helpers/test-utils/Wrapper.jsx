import { EditorState } from 'draft-js';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import TextEditor from '../../TextEditor';

const Wrapper = ({ error, helperText }) => {
  const [editorState, setEditorState] = useState(EditorState.createEmpty());

  return (
    <TextEditor
      error={error}
      helperText={helperText}
      editorState={editorState}
      setEditorState={setEditorState}
    />
  );
};

Wrapper.defaultProps = {
  error: false,
  helperText: null,
};

Wrapper.propTypes = {
  error: PropTypes.bool,
  helperText: PropTypes.string,
};

export default Wrapper;
