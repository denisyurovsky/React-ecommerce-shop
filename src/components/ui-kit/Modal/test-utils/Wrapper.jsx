import { Button } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { MODAL_SIZES } from '../../../../constants/constants';
import Modal from '../Modal';

const { NORMAL, SMALL } = MODAL_SIZES;

const Wrapper = ({
  title,
  actionButtonLabel,
  cancelButtonLabel,
  onConfirm,
  size,
}) => {
  const [isOpen, setIsOpen] = useState(false);

  const handleOpen = () => setIsOpen(true);
  const handleClose = () => setIsOpen(false);

  return (
    <>
      <Button onClick={handleOpen}>Modal</Button>
      <Modal
        isOpen={isOpen}
        onClose={handleClose}
        title={title}
        actionButtonLabel={actionButtonLabel}
        cancelButtonLabel={cancelButtonLabel}
        onConfirm={onConfirm}
        size={size}
      >
        <div>Content</div>
      </Modal>
    </>
  );
};

Wrapper.defaultProps = {
  title: null,
  actionButtonLabel: null,
  cancelButtonLabel: null,
  onConfirm: null,
};

Wrapper.propTypes = {
  title: PropTypes.string,
  actionButtonLabel: PropTypes.string,
  cancelButtonLabel: PropTypes.string,
  onConfirm: PropTypes.func,
  size: PropTypes.oneOf([NORMAL, SMALL]),
};

export default Wrapper;
