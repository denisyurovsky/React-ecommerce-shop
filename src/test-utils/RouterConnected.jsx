import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router, Route, Routes } from 'react-router-dom';

function RouterConnected({ component: Component, path, initialPaths }) {
  const history = createMemoryHistory({ initialEntries: initialPaths });

  return (
    <Router location={history.location} navigator={history}>
      <Routes>
        <Route path={path} element={Component} />
      </Routes>
    </Router>
  );
}

RouterConnected.propTypes = {
  component: PropTypes.object.isRequired,
  path: PropTypes.string,
  initialPaths: PropTypes.arrayOf(PropTypes.string),
};

RouterConnected.defaultProps = {
  path: '/',
  initialPaths: ['/'],
};

export default RouterConnected;
