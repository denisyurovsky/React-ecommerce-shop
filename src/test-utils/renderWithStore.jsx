import { configureStore } from '@reduxjs/toolkit';
import { render as rtlRender } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import cartReducer from '../store/cart/cartSlice';
import categoriesReducer from '../store/categories/categoriesSlice';
import feedbackReducer from '../store/feedback/feedbackSlice';
import productsReducer from '../store/products/productsSlice';
import userReducer from '../store/user/userSlice';

function render(component, { role = null, store } = {}) {
  let preloadedState;

  if (!store) {
    preloadedState = role
      ? {
          user: {
            user: {
              id: 5,
              role,
            },
          },
        }
      : {};
  }

  const storeWithState = store
    ? store
    : configureStore({
        reducer: {
          cart: cartReducer,
          products: productsReducer,
          categories: categoriesReducer,
          user: userReducer,
          feedback: feedbackReducer,
        },
        preloadedState,
      });

  function Wrapper({ children }) {
    return (
      <Provider store={storeWithState}>
        {children}
        <ToastContainer />
      </Provider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.object,
  };

  return rtlRender(component, { wrapper: Wrapper });
}

export * from '@testing-library/react';
export default render;
