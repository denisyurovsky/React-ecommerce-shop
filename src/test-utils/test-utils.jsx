import { configureStore } from '@reduxjs/toolkit';
import { render as rtlRender } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import { Provider } from 'react-redux';

import userReducer from '../store/user/userSlice';

const defaultStore = configureStore({ reducer: { user: userReducer } });

const render = (component, storage = defaultStore) => {
  const AllProviders = ({ children }) => (
    <Provider store={storage}>{children}</Provider>
  );

  AllProviders.propTypes = {
    children: PropTypes.node,
  };

  return rtlRender(component, { wrapper: AllProviders });
};

render.propTypes = {
  component: PropTypes.node.isRequired,
  storage: PropTypes.object,
};

export * from '@testing-library/react';
export default render;
