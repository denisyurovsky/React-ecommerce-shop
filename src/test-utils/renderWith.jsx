import { configureStore } from '@reduxjs/toolkit';
import { render as rtlRender } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import cartReducer from '../store/cart/cartSlice';
import categoriesReducer from '../store/categories/categoriesSlice';
import feedbackReducer from '../store/feedback/feedbackSlice';
import productsReducer from '../store/products/productsSlice';
import userReducer from '../store/user/userSlice';

import RouterConnected from './RouterConnected';

function render(
  component,
  preloadedState = {},
  path = '/',
  initialPaths = ['/']
) {
  const store = configureStore({
    reducer: {
      cart: cartReducer,
      products: productsReducer,
      categories: categoriesReducer,
      user: userReducer,
      feedback: feedbackReducer,
    },
    preloadedState,
  });

  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <RouterConnected
          component={children}
          path={path}
          initialPaths={initialPaths}
        />
        <ToastContainer />
      </Provider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.object,
  };

  return rtlRender(component, { wrapper: Wrapper });
}

export * from '@testing-library/react';
export default render;
