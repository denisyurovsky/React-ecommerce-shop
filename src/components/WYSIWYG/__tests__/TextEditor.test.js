import { render, fireEvent, screen, createEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import {
  EDITOR_MAX_SYMBOLS,
  MAX_LENGTH_MESSAGE,
  table,
  LINK_PREDEFINED_TEXT,
} from '../constants';
import { LONG_TEXT } from '../helpers/test-utils/testConstants';
import Wrapper from '../helpers/test-utils/Wrapper';

const { INPUT_ERROR, MAX_ROWS } = table;

window.scrollTo = jest.fn();

const getTextArea = () => screen.getByRole('textbox');
const pasteText = (text) =>
  createEvent.paste(getTextArea(), {
    clipboardData: {
      types: ['text/plain'],
      getData: () => text,
    },
  });

const addTableToEditor = () => {
  userEvent.click(screen.getByTestId('TableChartIcon'));
  fireEvent.change(screen.getByLabelText('Number of rows'), {
    target: { value: '2' },
  });
  userEvent.click(screen.getByRole('button', { name: /add table/i }));
};

describe('footer:', () => {
  describe('inner errors', () => {
    beforeEach(() => render(<Wrapper />));

    it('should show max size is reached message', async () => {
      await screen.findByText(/description/i);
      fireEvent(getTextArea(), pasteText(LONG_TEXT));

      expect(screen.getByText(MAX_LENGTH_MESSAGE)).toBeInTheDocument();
    });
  });

  describe('outer errors', () => {
    beforeEach(() => render(<Wrapper error={true} helperText="test error" />));

    it('should show outer error', () => {
      expect.assertions(2);
      const text = 'try';

      fireEvent(getTextArea(), pasteText(text));

      expect(screen.getByText(text)).toBeInTheDocument();
      expect(screen.getByText('test error')).toBeInTheDocument();
    });

    it('should display inner error with higher importance', () => {
      fireEvent(getTextArea(), pasteText(LONG_TEXT));

      expect(screen.getByText(MAX_LENGTH_MESSAGE)).toBeInTheDocument();
    });
  });

  describe('correct messages', () => {
    beforeEach(() => render(<Wrapper />));

    it('should show how many symbols are left to max value', () => {
      const text = 'try';
      const expectedLength = `${EDITOR_MAX_SYMBOLS - text.length}`;

      fireEvent(getTextArea(), pasteText(text));

      expect(screen.getByTestId('left-number')).toHaveTextContent(
        expectedLength
      );
    });

    it('should change styles for state with less than 50 symbols', () => {
      fireEvent(getTextArea(), pasteText(LONG_TEXT.slice(0, 990)));

      expect(screen.getByTestId('editor-footer')).toMatchSnapshot();
    });

    it('should show hint when anchor placed into table', () => {
      addTableToEditor();

      expect(
        screen.getByTestId('TipsAndUpdatesOutlinedIcon')
      ).toBeInTheDocument();
    });
  });
});

describe('modifiers functionality:', () => {
  describe('table', () => {
    beforeEach(() => render(<Wrapper />));

    it('should show error if entered number more than max table rows', async () => {
      await screen.findByText('Description');
      expect.assertions(2);

      userEvent.click(screen.getByTestId('TableChartIcon'));
      fireEvent.change(screen.getByLabelText('Number of rows'), {
        target: { value: MAX_ROWS + 2 },
      });

      expect(screen.getByRole('button', { name: /add table/i }).disabled).toBe(
        true
      );
      expect(screen.getByText(INPUT_ERROR)).toBeInTheDocument();
    });

    it('should create a new block after "Enter"', () => {
      fireEvent(getTextArea(), pasteText('block'));
      userEvent.keyboard('{enter}');
      fireEvent(getTextArea(), pasteText('one more block'));
      userEvent.keyboard('{enter}');

      addTableToEditor();

      userEvent.keyboard('{enter}');

      expect(screen.queryByTestId('TipsAndUpdatesOutlinedIcon')).toBeNull();
    });

    it('should create a new line after "Enter" + "Shift"', () => {
      fireEvent(getTextArea(), pasteText('block'));
      addTableToEditor();

      userEvent.keyboard('{shift}{enter}');

      expect(
        screen.queryByTestId('TipsAndUpdatesOutlinedIcon')
      ).toBeInTheDocument();
    });
  });

  describe('link', () => {
    beforeEach(() => render(<Wrapper />));

    it('should show predefined text for link', () => {
      userEvent.click(screen.getByTestId('AddLinkIcon'));

      expect(screen.getByLabelText(/insert hyperlink/i).value).toBe(
        LINK_PREDEFINED_TEXT
      );
    });

    it('should add correct link', () => {
      const displayedText = 'test link';

      userEvent.click(screen.getByTestId('AddLinkIcon'));
      fireEvent.change(screen.getByLabelText(/insert hyperlink/i), {
        target: { value: 'veryimportant.com' },
      });
      fireEvent.change(screen.getByLabelText(/displayed text/i), {
        target: { value: displayedText },
      });
      userEvent.click(screen.getByRole('button', { name: /add link/i }));

      expect(screen.getByText(displayedText)).toBeInTheDocument();
    });
  });

  describe('email', () => {
    beforeEach(() => render(<Wrapper />));

    it('should add correct email', () => {
      const email = 'test@email.com';

      userEvent.click(screen.getByTestId('MailIcon'));
      fireEvent.change(screen.getByLabelText(/e-mail/i), {
        target: { value: email },
      });
      userEvent.click(screen.getByRole('button', { name: /insert/i }));

      expect(screen.getByText(email)).toBeInTheDocument();
    });
  });

  describe('inline', () => {
    beforeEach(() => render(<Wrapper />));

    it('should apply styles', () => {
      const text = 'bold style';

      userEvent.click(screen.getByText(/total/i));
      userEvent.click(screen.getByTestId('FormatBoldIcon'));

      fireEvent(getTextArea(), pasteText(text));

      expect(screen.getByText(text).parentElement.style['font-weight']).toBe(
        'bold'
      );
    });
  });

  describe('block', () => {
    beforeEach(() => render(<Wrapper />));

    it('should apply list styles', () => {
      const listItem = 'first item';

      userEvent.click(screen.getByText(/total/i));
      userEvent.click(screen.getByTestId('FormatListNumberedIcon'));
      fireEvent(getTextArea(), pasteText(listItem));

      expect(screen.getByText(listItem).closest('ol')).toBeInTheDocument();
    });
  });
});

describe('general functionality', () => {
  beforeEach(() => render(<Wrapper />));

  it('should be able to react on "Backspace"', () => {
    const text = 'test343';

    fireEvent(getTextArea(), pasteText(text));

    userEvent.click(screen.getByText(text));
    userEvent.keyboard('{Backspace}{Backspace}{Backspace}');

    expect(screen.getByText(text.slice(0, 4))).toBeInTheDocument();
  });
});
