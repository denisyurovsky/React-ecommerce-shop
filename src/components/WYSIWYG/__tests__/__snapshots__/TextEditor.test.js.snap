// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`footer: correct messages should change styles for state with less than 50 symbols 1`] = `
<div
  class="footer MuiBox-root css-0"
  data-testid="editor-footer"
>
  <p
    class="MuiTypography-root MuiTypography-body1 symbols css-ahj2mt-MuiTypography-root"
  >
    Left: 
    <span
      class="warning"
      data-testid="left-number"
    >
      10
    </span>
  </p>
</div>
`;
